File: src/main/java/uk/edu/populationfitness/models/GenerationsAnalysis.java
package uk.edu.populationfitness.models;

public class GenerationsAnalysis {
    public final long maxExpectedPopulation;

    public final long historicalPopulationBeforeDisease;

    public final long historicalDiseaseTotalDeaths;

    public final long modernPopulationBeforeDisease;

    public final long modernDiseaseTotalDeaths;

    public final long totalBorn;

    public final long totalKilled;

    public GenerationsAnalysis(Iterable<GenerationStatistics> history) {
        long maxPop = 0;
        long born = 0;
        long killed = 0;
        long previousPopulation = 0;
        Epoch historicalDisease = null;
        Epoch modernDisease = null;
        long historicalPop = 0;
        long modernPop = 0;
        long historicalKilled = 0;
        long modernKilled = 0;
        for (GenerationStatistics generation: history) {
            maxPop = Math.max(maxPop, generation.epoch.expected_max_population);
            born += generation.number_born;
            killed += generation.number_killed;
            if (generation.epoch.disease()){
                if (historicalDisease == null){
                    historicalDisease = generation.epoch;
                    historicalPop = previousPopulation;
                    historicalKilled = 0;
                }
                if (historicalDisease.start_year == generation.epoch.start_year){
                    // Historical disease
                    historicalKilled += generation.number_killed;
                }
                else
                {
                    // Modern disease
                    if (modernDisease == null){
                        modernDisease = generation.epoch;
                        modernPop = previousPopulation;
                        modernKilled = 0;
                    }

                    if (modernDisease.start_year == generation.epoch.start_year){
                        // Modern disease
                        modernKilled += generation.number_killed;
                    }
                    else {
                        // Reset to later epoch
                        modernDisease = generation.epoch;
                        modernPop = previousPopulation;
                        modernKilled = generation.number_killed;
                    }
                }
            }

            previousPopulation = generation.population;
        }

        maxExpectedPopulation = maxPop;
        historicalPopulationBeforeDisease = historicalPop;
        historicalDiseaseTotalDeaths = historicalKilled;
        modernPopulationBeforeDisease = modernPop;
        modernDiseaseTotalDeaths = modernKilled;
        totalBorn = born;
        totalKilled = killed;
    }
}

File: Scala/PopulationFitness/src/main/scala/uk/edu/populationfitness/models/history/GenerationsAnalysis.scala
package uk.edu.populationfitness.models.history

class GenerationsAnalysis private {

  private var maxPop: Long = 0
  private var born: Long = 0
  private var killed: Long = 0
  private var previousPopulation: Long = 0
  private var historicalPop: Long = 0
  private var modernPop: Long = 0
  private var historicalKilled: Long = 0
  private var modernKilled: Long = 0

  private var historicalDisease: Option[Epoch] = None
  private var modernDisease: Option[Epoch] = None

  def maxExpectedPopulation = maxPop
  def historicalPopulationBeforeDisease = historicalPop
  def historicalDiseaseTotalDeaths = historicalKilled
  def modernPopulationBeforeDisease = modernPop
  def modernDiseaseTotalDeaths = modernKilled
  def totalBorn = born
  def totalKilled = killed

  def this(statistics: Seq[GenerationStatistics]) {
    this

    statistics.foldLeft(this)(_.add(_))
  }

  private def ensureHistoricalDefined(epoch: Epoch): Unit = {
    historicalDisease match {
      case None => {
        historicalDisease = Some(epoch)
        historicalPop = previousPopulation
        historicalKilled = 0
      }
      case _ =>
    }
  }

  private def ensureModernDefined(epoch: Epoch): Unit = {
    historicalDisease match {
      case None => {
        modernDisease = Some(epoch)
        modernPop = previousPopulation
        modernKilled = 0
      }
      case _ =>
    }
  }

  private def add(generation: GenerationStatistics): GenerationsAnalysis = {
    maxPop = Math.max(maxPop, generation.epoch.expectedMaxPopulation)
    born += generation.numberBorn
    killed += generation.numberKilled

    if (generation.epoch.disease) {
      ensureHistoricalDefined(generation.epoch)

      if (historicalDisease.get.startYear == generation.epoch.startYear) { // Historical disease
        historicalKilled += generation.numberKilled
      }
      else { // Modern disease
        ensureModernDefined(generation.epoch)

        if (modernDisease.get.startYear == generation.epoch.startYear)
          modernKilled += generation.numberKilled
        else { // Reset to later epoch
          modernDisease = Some(generation.epoch)
          modernPop = previousPopulation
          modernKilled = generation.numberKilled
        }
      }
    }
    previousPopulation = generation.population
    this
  }
}
